core-bundle/src/Controller/Page/AbstractPageController.php

```diff
@@ -0,0 +1,129 @@
+        $page->templateGroup = $layout->getRelated('pid')->templates ?? null;
+        $this->container->get('contao.image.picture_factory')->setDefaultDensities($layout->defaultImageDensities);
+        $this->container->get('contao.image.preview_factory')->setDefaultDensities($layout->defaultImageDensities);
+
+        // Create layout template and assign defaults
+        $template = $this->createTemplate($layout->template);
+        $this->addDefaultDataToTemplate($template, $page, $layout, $responseContext);
+
+        return $this->getResponse($template, $layout, $request);
+    }
+
+    public static function getSubscribedServices(): array
+    {
+        $services = parent::getSubscribedServices();
+
+        $services['contao.routing.response_context_factory'] = CoreResponseContextFactory::class;
+        $services['contao.security.token_checker'] = TokenChecker::class;
+        $services['contao.image.picture_factory'] = PictureFactoryInterface::class;
+        $services['contao.image.preview_factory'] = PreviewFactory::class;
+
+        return $services;
+    }
+
+    abstract protected function getResponse(LayoutTemplate $template, LayoutModel $model, Request $request): Response;
+
+    protected function createTemplate(string $identifier): LayoutTemplate
+    {
+        return new LayoutTemplate(
+            $identifier,
+            fn (LayoutTemplate $template, Response|null $preBuiltResponse): Response => $this->render("@Contao/{$template->getName()}.html.twig", $template->getData(), $preBuiltResponse),
+        );
+    }
+
+    protected function getResponseContext(PageModel $page): ResponseContext
+    {
+        // Initialize response context
+        return $this->container
+            ->get('contao.routing.response_context_factory')
+            ->createContaoWebpageResponseContext($page)
+        ;
+    }
+
+    protected function addDefaultDataToTemplate(LayoutTemplate $template, PageModel $page, LayoutModel $layout, ResponseContext $responseContext): void
+    {
+        // Slots
+        $moduleIdsBySlot = [];
+
+        foreach (StringUtil::deserialize($layout->modules, true) as $definition) {
+            if ($definition['enable']) {
+                $id = (int) $definition['mod'];
+                $moduleIdsBySlot[$definition['col']][] = $id;
+            }
+        }
+
+        foreach ($moduleIdsBySlot as $slot => $moduleIds) {
+            $template->setSlot($slot, $this->composeModulesForSlot($slot, $moduleIds));
+        }
+
+        // Chrome
+        $template->set('page', $page);
+        $template->set('layout', $layout);
+
+        $template->set('head', $responseContext->get(HtmlHeadBag::class));
+        $template->set('preview_mode', $this->container->get('contao.security.token_checker')->isPreviewMode());
+
+        $locale = LocaleUtil::formatAsLocale($page->language);
+        $isRtl = (\ResourceBundle::create($locale, 'ICUDATA')['layout']['characters'] ?? null) === 'right-to-left';
```

Comment:
Smartass! If that's correct (cc @ausi) I would propose adding this to our `LocaleUtil`.

Reply:
Looks correct 😎
+1 for locale util